Split ConcreteModuleType into two types

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29826 Fix type sharing on loaded ScriptModules
* #29825 Re-unify module and interface in ConcreteModuleType
* **#29824 Split ConcreteModuleType into two types**

We have two distinct phases/uses for ConcreteModuleType:
1. We are building it up and using it to check whether we can
reuse JIT types. (RawConcreteModuleType)
2. We are using it to satisfy ModuleValue::attr queries.
(ConcreteModuleType)

These types share an underlying `ConcreteModuleTypeData` which 
actually stores the relevant info.

Previously they were the same type because I was lazy, but it's been the
source of a bug. So split them to formalize the differing invariants for
the two phases.

Differential Revision: [D18575010](https://our.internmc.facebook.com/intern/diff/D18575010)

suo committed:
Split ConcreteModuleType into two types

We have two distinct phases/uses for ConcreteModuleType:
1. We are building it up and using it to check whether we can
reuse JIT types. (RawConcreteModuleType)
2. We are using it to satisfy ModuleValue::attr queries.
(ConcreteModuleType)

Previously they were the same type because I was lazy, but it's been the
source of a bug. So split them to formalize the differing invariants for
the two phases.

suo committed:
Update on "Split ConcreteModuleType into two types"


We have two distinct phases/uses for ConcreteModuleType:
1. We are building it up and using it to check whether we can
reuse JIT types. (RawConcreteModuleType)
2. We are using it to satisfy ModuleValue::attr queries.
(ConcreteModuleType)

These types share an underlying `ConcreteModuleTypeData` which 
actually stores the relevant info.

Previously they were the same type because I was lazy, but it's been the
source of a bug. So split them to formalize the differing invariants for
the two phases.

zdevito commented on torch/csrc/jit/script/concrete_module_type.h:
```diff
@@ -41,83 +43,24 @@ enum class IterableModuleKind { NONE, LIST, DICT };
 
+// ConcreteModuleType has two phases.
+// 1. Creation: First we build it up, during the ScriptModule conversion
+// process. This is represented by RawConcreteModuleType.
```
Before I found this comment, but after reading the diff I could not figure out which of the two types was which. Can we call RawConcreteModuleType ConcreteModuleTypeBuilder?

suo replied:
lol that was my first name for it 😛

zdevito commented on torch/csrc/jit/script/concrete_module_type.h:
```diff
@@ -256,11 +184,92 @@ class VISIBILITY_HIDDEN ConcreteModuleType {
+
+// Represents a concrete type during in the process for construction. We use
+// this to decide whether we can share types between modules.
+class VISIBILITY_HIDDEN RawConcreteModuleType {
```
If ConcreteModuleType was a friend of its builder, then we would not need ConcreteModuleTypeData and its data_ indirections. I don't think that changes its safety properties as long as the builder returns the built type by value to cause it to create a copy.

zdevito reviewed (approved):
Minor nits, but this otherwise looks good.

suo committed:
Update on "Split ConcreteModuleType into two types"


We have two distinct phases/uses for ConcreteModuleType:
1. We are building it up and using it to check whether we can
reuse JIT types. (RawConcreteModuleType)
2. We are using it to satisfy ModuleValue::attr queries.
(ConcreteModuleType)

These types share an underlying `ConcreteModuleTypeData` which 
actually stores the relevant info.

Previously they were the same type because I was lazy, but it's been the
source of a bug. So split them to formalize the differing invariants for
the two phases.

Differential Revision: [D18575010](https://our.internmc.facebook.com/intern/diff/D18575010)

suo committed:
Update on "Split ConcreteModuleType into two types"


We have two distinct phases/uses for ConcreteModuleType:
1. We are building it up and using it to check whether we can
reuse JIT types. (RawConcreteModuleType)
2. We are using it to satisfy ModuleValue::attr queries.
(ConcreteModuleType)

These types share an underlying `ConcreteModuleTypeData` which 
actually stores the relevant info.

Previously they were the same type because I was lazy, but it's been the
source of a bug. So split them to formalize the differing invariants for
the two phases.

Differential Revision: [D18575010](https://our.internmc.facebook.com/intern/diff/D18575010)
